lib/airbrake/rack/instrumentable.rb

```diff
@@ -15,12 +17,99 @@ module Rack
     # @since v9.2.0
     module Instrumentable
       def airbrake_capture_timing(method_name, label: method_name.to_s)
-        alias_method "#{method_name}_without_airbrake", method_name
+        mod = ::Airbrake::Rack::Instrumentable
+        if mod.should_prepend?(self, method_name)
+          mod.prepend_airbrake_capture_timing(self, method_name, label)
+        else
+          mod.chain_airbrake_capture_timing(self, method_name, label)
+        end
+        method_name
+      end
+
+      # @api private
+      def __airbrake_capture_timing_module__
+        # Module used to store prepended wrapper methods, saved as an instance
+        # variable so each target class/module gets its own module. This just
+        # a convenience to avoid prepending a lot of anonymous modules.
+        @__airbrake_capture_timing_module__ ||= ::Module.new
+      end
+
+      # Using api private self methods so they don't get defined in the target
+      # class or module, but can still be called by the above method.
+
+      # @api private
+      def self.should_prepend?(klass, method_name)
+        # Don't chain already-prepended or operator methods.
+        klass.module_exec do
+          self_class_idx = ancestors.index(self)
+          method_owner_idx = ancestors.index(instance_method(method_name).owner)
+          method_owner_idx < self_class_idx || !(/\A\W/ =~ method_name).nil?
+        end
+      end
 
-        define_method(method_name) do |*args|
-          Airbrake::Rack.capture_timing(label) do
-            __send__("#{method_name}_without_airbrake", *args)
+      # @api private
+      def self.prepend_airbrake_capture_timing(klass, method_name, label)
+        args = method_signature
+        klass.module_exec do
+          # Check the visibility.
```

Comment:
I don't like how we repeat this code for both cases so I gave this a shot and factored out it into a separate method. 
* No need to define 3 temp variables, just 1 now
* We declare visibility directly when evalling

<details>
<summary>click to see the diff</summary>

```diff
diff --git a/lib/airbrake/rack/instrumentable.rb b/lib/airbrake/rack/instrumentable.rb
index 8d6b349..da0c702 100644
--- a/lib/airbrake/rack/instrumentable.rb
+++ b/lib/airbrake/rack/instrumentable.rb
@@ -50,10 +50,9 @@ module Airbrake
       # @api private
       def self.prepend_airbrake_capture_timing(klass, method_name, label)
         args = method_signature
+        visibility = check_visibility(klass, method_name)
+
         klass.module_exec do
-          # Check the visibility.
-          is_protected = protected_method_defined?(method_name)
-          is_private = private_method_defined?(method_name)
           mod = __airbrake_capture_timing_module__
           mod.module_exec do
             # Generate the wrapper method.
@@ -63,10 +62,8 @@ module Airbrake
                   super
                 end
               end
+              #{visibility} :#{method_name}
             RUBY
-            # Restore the visibility.
-            protected method_name if is_protected
-            private method_name if is_private
           end
           prepend mod
         end
@@ -75,29 +72,37 @@ module Airbrake
       # @api private
       def self.chain_airbrake_capture_timing(klass, method_name, label)
         args = method_signature
-        klass.module_exec do
-          # Check the visibility.
-          is_public = private_method_defined?(method_name)
-          is_protected = protected_method_defined?(method_name)
-          is_private = private_method_defined?(method_name)
+        visibility = check_visibility(klass, method_name)
+
+        # Generate the wrapper method.
+        aliased = method_name.to_s.sub(/([?!=])$/, '')
+        punctuation = Regexp.last_match(1)
+        wrapped_method_name = "#{aliased}_without_airbrake#{punctuation}"

-          # Generate the wrapper method.
-          aliased = method_name.to_s.sub(/([?!=])$/, '')
-          punctuation = Regexp.last_match(1)
-          alias_method "#{aliased}_without_airbrake#{punctuation}", method_name
+        klass.module_exec do
+          alias_method wrapped_method_name, method_name
           remove_method method_name
           module_eval <<-RUBY, __FILE__, __LINE__ + 1
             def #{method_name}(#{args})
               Airbrake::Rack.capture_timing(#{label.to_s.inspect}) do
-                __send__("#{aliased}_without_airbrake#{punctuation}", #{args})
+                __send__(:#{wrapped_method_name}, #{args})
               end
             end
+            #{visibility} :#{method_name}
           RUBY
+        end
+      end

-          # Restore the visibility.
-          public method_name if is_public
-          protected method_name if is_protected
-          private method_name if is_private
+      # @api private
+      def self.check_visibility(klass, method_name)
+        klass.module_exec do
+          if protected_method_defined?(method_name)
+            'protected'
+          elsif private_method_defined?(method_name)
+            'private'
+          else
+            'public'
+          end
         end
       end
</details>

Reply:
I like it.